service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -82,107 +126,147 @@ public ResponseEntity<RunSetStateResponse> postRunSet(RunSetRequest request) {
     }
 
-    // Store the run:
+    return Optional.empty();
+  }
+
+  private RunStateResponse storeRun(
+      UUID runId,
+      String externalId,
+      RunSet runSet,
+      String recordId,
+      CbasRunStatus runState,
+      String errors) {
+    String additionalErrorMsg = "";
     int created =
         runDao.createRun(
             new Run(
                 runId,
-                workflowResponse.getRunId(),
+                externalId,
                 runSet,
-                dataTableRowId,
+                recordId,
                 OffsetDateTime.now(),
-                UNKNOWN,
+                runState,
                 OffsetDateTime.now(),
                 OffsetDateTime.now(),
-                null));
+                errors));
 
     if (created != 1) {
-      log.error(
-          "New workflow for record {} submitted to engine as {} but CBAS database INSERT returned '{} rows created'",
-          dataTableRowId,
-          workflowResponse.getRunId(),
-          created);
+      additionalErrorMsg =
+          String.format(
+              "CBAS failed to create new row for Record ID %s in %s state in database. INSERT returned '%s rows created'",
+              recordId, runState, created);
+      log.error(additionalErrorMsg);
     }
 
-    // Return the result:
-    return new ResponseEntity<>(
-        new RunSetStateResponse()
-            .runSetId(runSetId.toString())
-            .addRunsItem(
-                new RunStateResponse()
-                    .runId(runId.toString())
-                    .state(CbasRunStatus.toCbasApiState(UNKNOWN)))
-            .state(RunSetState.RUNNING),
-        HttpStatus.OK);
+    return new RunStateResponse()
+        .runId(runId.toString())
+        .state(CbasRunStatus.toCbasApiState(runState))
+        .errors(errors + additionalErrorMsg);
   }
 
-  public static boolean requestIsValid(RunSetRequest request, CbasApiConfiguration config) {
-    int recordIdsSize = request.getWdsRecords().getRecordIds().size();
-    int recordIdsMax = config.getRunSetsMaximumRecordIds();
-    if (recordIdsSize > recordIdsMax) {
-      log.warn(
-          "Bad user request: %s record IDs submitted exceeds the maximum value of %s"
-              .formatted(recordIdsSize, recordIdsMax));
-      return false;
+  private List<RunStateResponse> buildInputsAndSubmitRun(
+      RunSetRequest request, RunSet runSet, ArrayList<RecordResponse> recordResponses) {
+    ArrayList<RunStateResponse> runStateResponseList = new ArrayList<>();
+
+    for (RecordResponse record : recordResponses) {
+      // Build the inputs set from workflow parameter definitions and the fetched record
+      Map<String, Object> workflowInputs =
+          InputGenerator.buildInputs(request.getWorkflowInputDefinitions(), record);
+
+      // Submit the workflow, get its ID and store the Run to database
+      RunId workflowResponse;
+      UUID runId = UUID.randomUUID();
```

Comment:
TOL: In some future version, we'll probably want to store this first, separately from trying to actually submit (in case we get restarted an at inopportune time, for example)